src/Features/Core/Portable/NavigateTo/AbstractNavigateToSearchService.cs

```diff
@@ -61,11 +61,27 @@ internal abstract partial class AbstractNavigateToSearchService : INavigateToSea
                     (service, solutionInfo, cancellationToken) => service.SearchProjectAsync(solutionInfo, project.Id, priorityDocumentIds, searchPattern, kinds.ToImmutableArray(), cancellationToken),
                     cancellationToken).ConfigureAwait(false);
 
-                return result.HasValue ? result.Value.SelectAsArray(r => r.Rehydrate(solution)) : ImmutableArray<INavigateToSearchResult>.Empty;
+                return await RehydrateAsync(result, solution, cancellationToken).ConfigureAwait(false);
             }
 
             return await SearchProjectInCurrentProcessAsync(
                 project, priorityDocuments, searchPattern, kinds, cancellationToken).ConfigureAwait(false);
         }
+
+        private static async Task<ImmutableArray<INavigateToSearchResult>> RehydrateAsync(
```

Comment:
💡 Might as well avoid an unnecessary `Task<T>` allocation for the empty/synchronous return case:

```suggestion
        private static async ValueTask<ImmutableArray<INavigateToSearchResult>> RehydrateAsync(
```